sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -189,7 +189,15 @@ case class DataSource(
       throw new AnalysisException(
         s"Unable to infer schema for $format. It must be specified manually.")
     }
-    (dataSchema, partitionSchema)
+
+    // Override the fields of the partition schema if the data schema has the same field
+    val resolvedPartitionSchema = partitionSchema.map { partitionField =>
```

Comment:
I don't think we need this. Otherwise, the value you return when `if (justPartitioning)` is inconsistent.
In a real world setting, as you provided in your example, for:
`case class A(a: Long, b: Int)`
if `a` is in fact a `Long` and is also the partitioning column, there will be examples of `a` being a `Long` among the partition columns, therefore, things should work. No-one should ever have more than 2 billion partition columns anyway. Spark wouldn't be able to resolve that many columns all in-memory right now anyway